x/market/keeper/keeper.go

```diff
@@ -417,6 +419,61 @@ func (k *Keeper) CancelReplaceOrder(ctx sdk.Context, newOrder types.Order, origC
 	return &sdk.Result{Events: evts}, nil
 }
 
+func (k *Keeper) CancelReplaceMarketOrder(ctx sdk.Context, msg *types.MsgCancelReplaceMarketOrder) (*sdk.Result, error) {
```

Comment:
The pattern is usually to not send `Msg` objects into the Keeper, but I can see why it's necessary in this case. There are so many parameters that need to be handled.

Could you evaluate the following approach, and see if it results in a better structure and avoids the `Msg` object in the Keeper:
1. Change `NewMarketOrderWithSlippage` to NOT submit the order to the book, but merely calculate the price and return a new `Order` instance.
2. Use that function in the message handler to create an Order object that can then be submitted to `CancelReplaceMarketOrder`.

It's possible that I overlooked something, and the above is untenable. :)

Reply:
To keep the Keeper API symmetric the same principle can be applied to the AddLimit / MarketOrder and will reuse the same price formula. Shall I?

Reply:
Yes, I think it should probably apply to all cases.